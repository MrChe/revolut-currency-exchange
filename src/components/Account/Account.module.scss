.Account {
  position: relative;
  width: 100%;
}
