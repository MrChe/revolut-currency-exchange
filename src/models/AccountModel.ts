import { action, makeAutoObservable, observable } from "mobx";
import { v4 as uuidv4 } from "uuid";
// import { persistStore } from "../utils/mobx-persist.utils";

export class AccountModel {
  public id: string;
  public currency: string;
  public balance: number;
  // public history: [];
  constructor(data: Partial<AccountModel>) {
    this.id = data.currency || uuidv4();
    this.currency = data.currency || "";
    this.balance = Number(data.balance?.toFixed(2)) || 100;
    // this.history = data.history;
    makeAutoObservable(this, {
      currency: observable,
      balance: observable,
      updateBalance: action,
    });
    // persistStore(this, ["id", "currency", "balance"], "AccountModel");
  }

  public updateBalance = (balance: number): void => {
    this.balance = Number(balance.toFixed(2));
  };
}
