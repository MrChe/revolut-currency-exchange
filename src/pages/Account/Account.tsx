import React from 'react';

export const Account = (): JSX.Element => {
  return <div>Account</div>;
};
