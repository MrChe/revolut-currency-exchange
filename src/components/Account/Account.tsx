import React from "react";
// import { observer } from "mobx-react-lite";
import cn from "classnames";
import styles from "./Account.module.scss";
// import { useStore } from "../../models/connect";
import { AccountModel } from "../../models/AccountModel";

interface IAccountProps {
  // type: "from" | "to";
  view: "preview" | "exchange";
  currency: string;
  currencyName: string;
  balance: string;
}

export const Account = (props: IAccountProps): JSX.Element => {
  const { currency, currencyName, balance } = props;
  return (
    <div
      className={cn({
        [styles.Account]: true,
        [styles.View_Preview]: props.view === "preview",
        [styles.View_Exchange]: props.view === "exchange",
      })}
    >
      <div className={styles.CurrencyInformation}>
        <div className={styles.Wrapper}>
          {props.view === "exchange" ? (
            <div>
              <p className={styles.CurrencyTitle}>
                {
                  // activeAccount.currency
                  currency
                }
              </p>
              <p className={styles.CurrencyText}>
                You have: {balance}
                {/*{ExchangeModel.formatCurrency(*/}
                {/*  activeAccount.balance,*/}
                {/*  activeAccount.currency,*/}
                {/*)}*/}
              </p>
            </div>
          ) : (
            <div>
              <p className={styles.CurrencyTitle}>
                {balance}
                {/*{ExchangeModel.formatCurrency(*/}
                {/*  activeAccount.balance,*/}
                {/*  activeAccount.currency,*/}
                {/*)}*/}
              </p>
              <p>
                {
                  currencyName
                  // currencyNames[activeAccount.currency]
                }
              </p>
            </div>
          )}
        </div>
      </div>
    </div>
  );
};
