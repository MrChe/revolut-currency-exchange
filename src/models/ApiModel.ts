import Axios, { AxiosInstance } from "axios";
import { IRates } from "../models/ExchangeModel";
import { filterRates } from "../utils/currencies";

const APP_ID = "e285ab8ceb004923b3e423a0d0583915";

export class ApiModel {
  private client: AxiosInstance;
  constructor() {
    this.client = Axios.create({
      baseURL: `https://openexchangerates.org/api`,
      params: {
        base: "USD",
        app_id: APP_ID,
        symbols: "USD,EUR,GBP,UAH",
      },
      headers: {
        "Content-Type": "application/json",
      },
    });
  }

  public getLatestRatesRequest = async (): Promise<IRates | undefined> => {
    try {
      const response = await this.client.get(`/latest.json`);
      if (!response.data) {
        throw new Error("Not have any data");
      }
      return response.data;
    } catch (error) {
      console.error("getLatestRatesRequest", error);
    }
  };
}
