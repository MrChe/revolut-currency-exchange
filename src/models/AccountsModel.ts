import { RootModel } from "./RootModel";
// import { persistStore } from "../utils/mobx-persist.utils";
import { action, computed, makeAutoObservable, observable } from "mobx";
import { AccountModel } from "./AccountModel";
import { IRates } from "./ExchangeModel";

export class AccountsModel {
  rootModel: RootModel;
  public accounts: AccountModel[];
  public ratesData: IRates | null;
  public selectedAccount: AccountModel | null;
  constructor(rootModel: RootModel, currencies: string[], base: string) {
    this.rootModel = rootModel;
    this.selectedAccount = null;
    this.accounts = [];
    this.ratesData = null;
    makeAutoObservable(this, {
      // rootModel: false,
      accounts: observable,
      selectedAccount: observable,
      accountsAsArray: computed,
      setSelectedAccount: action,
    });

    // INIT Module
    this.init(currencies, base);
    // persistStore(
    //   this,
    //   ["accounts", "selectedAccount", "accountsAsArray"],
    //   "AccountsModel",
    // );
  }

  private init = async (currencies: string[], base: string): Promise<void> => {
    try {
      if (currencies.length !== 0) {
        const data = await this.rootModel.ApiModel.getLatestRatesRequest(
          currencies,
          base,
        );
        if (data) {
          this.setRatesData(data);
          this.setAccounts(data);
        }
      } else {
        throw new Error("Please add some currencies like ['USD', 'UAH']");
      }
    } catch (error) {
      console.error("init", error);
    }
  };

  public setSelectedAccount = (id: string): void => {
    const foundAccount = this.getAccountById(id);
    if (foundAccount) {
      this.selectedAccount = foundAccount;
    } else {
      console.error(`Cant find account by id: ${id}`);
    }
  };

  public get accountsAsArray(): AccountModel[] {
    return this.accounts.slice();
  }

  public setRatesData = (ratesData: IRates): void => {
    this.ratesData = ratesData;
  };

  public setAccounts = (ratesData: IRates): void => {
    Object.keys(ratesData.rates).forEach((currency: string) => {
      this.accounts.push(
        new AccountModel({
          currency,
          balance: 100,
        }),
      );
    });
  };

  public getAccountById = (id: string): AccountModel | undefined => {
    return this.accounts.find((account) => account.id === id);
  };
}
