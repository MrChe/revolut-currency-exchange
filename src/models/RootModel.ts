import { ExchangeModel } from "./ExchangeModel";
import { AccountsModel } from "./AccountsModel";
import { ApiModel } from "./ApiModel";

export class RootModel {
  public ExchangeModel: ExchangeModel;
  public AccountsModel: AccountsModel;
  public ApiModel: ApiModel;
  constructor() {
    this.ApiModel = new ApiModel();
    // init accounts
    this.AccountsModel = new AccountsModel(
      this,
      ["USD", "EUR", "GBP", "UAH"],
      "USD",
    );
    this.ExchangeModel = new ExchangeModel(this);
  }
}

export const rootModel = new RootModel();
