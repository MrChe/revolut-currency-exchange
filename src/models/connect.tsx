// holds a reference to the store (singleton)
import { createContext, useContext } from "react";
import { RootModel } from "./RootModel";

// create the context
const MobxContext = createContext<RootModel | null>(null);

// create the provider component
export const MobxProvider = MobxContext.Provider;

// create the hook
export const useStore = (): RootModel => {
  const context = useContext(MobxContext);
  if (!context) {
    throw new Error("useStore must be used within MobxContext");
  }

  return context;
};
