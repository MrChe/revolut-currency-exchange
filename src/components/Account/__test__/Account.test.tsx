import React, { Fragment } from "react";
import renderer from "react-test-renderer";
import { rootModel } from "../../../models/RootModel";
import { Account } from "../Account";
import { TestWrapper } from "../../../../tests/testWrapper";
import { ratesData } from "../../../../tests/mockData";
import { formatCurrency, currencyNames } from "../../../utils/helpers";

describe("Test Account Component", () => {
  const ExchangeModel = rootModel.ExchangeModel;
  beforeAll(() => {
    ExchangeModel.setRatesData(ratesData);
    ExchangeModel.setAccounts(ratesData);
    ExchangeModel.setActiveAccountFrom(ExchangeModel.accounts[0].id);
    ExchangeModel.setActiveAccountTo(ExchangeModel.accounts[1].id);
  });

  it("Account type Preview snapshot", () => {
    const tree = renderer
      .create(
        <TestWrapper>
          {ExchangeModel.activeAccountFrom ? (
            <Account
              balance={formatCurrency(
                ExchangeModel.activeAccountFrom?.balance,
                ExchangeModel.activeAccountFrom?.currency,
              )}
              view={"preview"}
              currency={ExchangeModel.activeAccountFrom?.currency}
              currencyName={
                currencyNames[ExchangeModel.activeAccountFrom?.currency] || ""
              }
            />
          ) : (
            <Fragment />
          )}
        </TestWrapper>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Account type Exchange snapshot", () => {
    const tree = renderer
      .create(
        <TestWrapper>
          {ExchangeModel.activeAccountFrom ? (
            <Account
              balance={formatCurrency(
                ExchangeModel.activeAccountFrom?.balance,
                ExchangeModel.activeAccountFrom?.currency,
              )}
              view={"exchange"}
              currency={ExchangeModel.activeAccountFrom?.currency}
              currencyName={
                currencyNames[ExchangeModel.activeAccountFrom?.currency] || ""
              }
            />
          ) : (
            <Fragment />
          )}
        </TestWrapper>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
