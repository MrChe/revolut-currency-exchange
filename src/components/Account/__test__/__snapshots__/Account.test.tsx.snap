// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Account Component Account type Exchange snapshot 1`] = `
<div
  className="Account View_Exchange"
>
  <div
    className="CurrencyInformation"
  >
    <div
      className="Wrapper"
    >
      <div>
        <p
          className="CurrencyTitle"
        >
          EUR
        </p>
        <p
          className="CurrencyText"
        >
          You have: 
          €100.00
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Test Account Component Account type Preview snapshot 1`] = `
<div
  className="Account View_Preview"
>
  <div
    className="CurrencyInformation"
  >
    <div
      className="Wrapper"
    >
      <div>
        <p
          className="CurrencyTitle"
        >
          €100.00
        </p>
        <p>
          Euro
        </p>
      </div>
    </div>
  </div>
</div>
`;
